street_network/src/transform/separate_cycletracks.rs

```diff
@@ -0,0 +1,215 @@
+    if cycleway_lanes[0].lt == LaneType::Shoulder {
+        cycleway_lanes.remove(0);
+    }
+    if cycleway_lanes.last().as_ref().unwrap().lt == LaneType::Shoulder {
+        cycleway_lanes.pop();
+    }
+
+    // The cycleway was tagged as a separate way due to some kind of physical separation. We'll
+    // represent that with a buffer lane.
+    let buffer = LaneSpec {
+        // TODO Use https://wiki.openstreetmap.org/wiki/Proposed_features/cycleway:separation if
+        // available
+        lt: LaneType::Buffer(BufferType::Planters),
+        dir: Direction::Fwd,
+        width: LaneSpec::typical_lane_width(LaneType::Buffer(BufferType::Planters)),
+    };
+
+    // For every main road segment corresponding to the cycleway, we need to insert these
+    // cycleway_lanes somewhere.
+    //
+    // - Fixing the direction of the lanes
+    // - Appending them on the left or right side (and "inside" the inferred sidewalk on the road)
+    // - Inserting the buffer
+    for r in input.main_roads {
+        let main_road = &streets.roads[&r];
+        // Which side is closer to the cycleway?
+        let (left, right) = main_road.get_untrimmed_sides().unwrap();
+        // TODO georust has a way to check distance of linestrings. But for now, just check the
+        // middles
+        let snap_to_left = input.cycleway_center.middle().dist_to(left.middle())
+            < input.cycleway_center.middle().dist_to(right.middle());
+
+        // Does the cycleway point the same direction as this main road? We can use the left or
+        // right side, doesn't matter.
+        // TODO Check this logic very carefully; angles always lead to bugs. 90 is a very generous
+        // definition of parallel. But we have a binary decision to make, so maybe we should even
+        // use 180.
+        let oriented_same_way = input
+            .cycleway_center
+            .overall_angle()
+            .approx_eq(left.overall_angle(), 90.0);
+
+        // Where should we insert the cycleway lanes? If the main road already has a sidewalk,
+        // let's assume it should stay at the outermost part of the road. (That isn't always true,
+        // but it's an assumption we'll take for now.)
+        let insert_idx = if snap_to_left {
+            if main_road.lane_specs_ltr[0].lt.is_walkable() {
+                1
+            } else {
+                0
+            }
+        } else {
+            if main_road
+                .lane_specs_ltr
+                .last()
+                .as_ref()
+                .unwrap()
+                .lt
+                .is_walkable()
+            {
+                main_road.lane_specs_ltr.len() - 1
+            } else {
+                main_road.lane_specs_ltr.len()
+            }
+        };
+
+        streets.debug_road(r, format!("snap_to_left = {snap_to_left}, oriented_same_way = {oriented_same_way}, insert_idx = {insert_idx}"));
+
+        // This logic thankfully doesn't depend on driving side at all!
```

Comment:
I struggled to write this without scribbling diagrams on paper and using temporary debug visualization. It was a similar experience to [sausage links](https://github.com/a-b-street/osm2streets/blob/606cf03f46e79f69ea91a1cd9d62b9db140f2c07/street_network/src/transform/sausage_links.rs#L96). I hope there will be some better patterns to reason through this kind of thing. Maybe #1 is relevant.